draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+
+  @task
+  def foo():
+    with retry(max_retries=10, interval_start=0, interval_step=5, interval_max=120):
+      first_operation()
+
+    with retry(max_retries=10, interval_start=0, interval_step=5, interval_max=120):
+      second_operation()
+
+By default messages which cannot be re-published will be stored
+in the :ref:`draft/celery-5-high-level-architecture:messages backlog`.
+
+Implementers may provide other fallbacks such as executing the retried task
+in the same worker or abandoning the task entirely.
+
+Health Checks
+~~~~~~~~~~~~~
+
+Health Checks are used in Celery to verify that a worker is able to
+successfully execute a :ref:`task <draft/celery-5-high-level-architecture:Tasks>`
+or a :ref:`service <draft/celery-5-high-level-architecture:Services>`.
+
+The :ref:`draft/celery-5-high-level-architecture:Scheduler` is responsible
+for scheduling the health checks for execution in each worker after
+each time the configured period of time lapses.
+
+Whenever a health check should be executed the
+:ref:`draft/celery-5-high-level-architecture:Scheduler` instructs the
+:ref:`draft/celery-5-high-level-architecture:Publisher` to send the
+`<health check name>_expired` :term:`Event Message` to each worker's
+:ref:`draft/celery-5-high-level-architecture:Inbox Queue`.
+
+Workers which have tasks subscribed to the event will
+execute all the subscribed tasks in order to determine the state of the
+health check.
+
+Health Checks can handle :term:`Document Messages <Document Message>` as input
+from :ref:`draft/celery-5-high-level-architecture:Ingress Only Data Sources`.
+
+This is useful when you want to respond to an alert from a monitoring system
+or when you want to verify that all incoming data from said source is
+valid at all times before executing the task.
+
+In addition to tasks, Health Checks can also use
+:ref:`draft/celery-5-high-level-architecture:Services` in order to track
+changes in the environment it is running on.
+
+.. admonition:: Example
+
+  We have a task which requires 8GB of memory to complete.
+  The worker runs a service which constantly monitors the system's available
+  memory.
+  If there is not enough memory it changes the task's health check to the
+  **Unhealthy** state.
+
+If a task or a service that is part of a health check fails unexpectedly it
+is ignored and an error message is logged.
+
+Celery provides many types of health checks in order to verify that it can
+operate without any issues.
+
+Users may implement their own health checks in addition to the built-in health
+checks.
+
+Some health checks are specific to the worker they are executing on.
+Therefore, their state is stored in-memory in the worker.
+
+Other health checks are global to all or a group of workers.
+As such, their state is stored externally.
+
+If the state storage for health checks is not provided, these health checks
+are disabled.
+
+Health Checks can be associated with tasks in order to ensure that they are
```

Comment:
So the Health Check API, should mark tasks as successful and maintain the count of tasks considered completed or not (regardless of raised exceptions)?